infrastructure/src/main/java/org/corfudb/infrastructure/RequestHandlerMethods.java

```diff
@@ -0,0 +1,164 @@
+     * on another thread.
+     */
+    @FunctionalInterface
+    public interface HandlerMethod {
+        void handle(@Nonnull RequestMsg req,
+                    @Nonnull ChannelHandlerContext ctx,
+                    @Nonnull IServerRouter r);
+    }
+
+    /**
+     * Get the types of requests this handler will handle.
+     *
+     * @return  A set containing the types of requests this handler will handle.
+     */
+    public Set<RequestPayloadMsg.PayloadCase> getHandledTypes() {
+        return handlerMap.keySet();
+    }
+
+    /**
+     * Construct a new instance of RequestHandlerMethods.
+     */
+    public RequestHandlerMethods() {
+        handlerMap = new EnumMap<>(RequestPayloadMsg.PayloadCase.class);
+    }
+
+    /**
+     * Handle an incoming Corfu request message.
+     *
+     * @param req       The request message to handle.
+     * @param ctx       The channel handler context.
+     * @param r         The server router.
+     */
+    @SuppressWarnings("unchecked")
+    public void handle(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        final HandlerMethod handler = handlerMap.get(req.getPayload().getPayloadCase());
+        try {
+            handler.handle(req, ctx, r);
+        } catch (Exception e) {
+            log.error("handle[{}]: Unhandled exception processing {} request",
+                    req.getHeader().getRequestId(), req.getPayload().getPayloadCase(), e);
+
+            HeaderMsg responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getUnknownErrorMsg(e)), ctx);
+        }
+    }
+
+    /**
+     * Generate handlers for a particular server.
+     *
+     * @param caller    The context that is being used. Call MethodHandles.lookup() to obtain.
+     * @param server    The object that implements the server.
+     * @return          New request handlers for caller class.
+     */
+    public static RequestHandlerMethods generateHandler(@Nonnull final MethodHandles.Lookup caller,
+                                                        @NonNull final AbstractServer server) {
+        RequestHandlerMethods handler = new RequestHandlerMethods();
+        Arrays.stream(server.getClass().getDeclaredMethods())
+                .filter(method -> method.isAnnotationPresent(RequestHandler.class))
+                .forEach(method -> handler.registerMethod(caller, server, method));
+        return handler;
+    }
+
+    private void registerMethod(@Nonnull final MethodHandles.Lookup caller,
+                                @Nonnull final AbstractServer server,
+                                @Nonnull final Method method) {
+        final RequestHandler annotation = method.getAnnotation(RequestHandler.class);
+
+        if (handlerMap.containsKey(annotation.type())) {
+            throw new UnrecoverableCorfuError("HandlerMethod for " + annotation.type() + " already registered!");
```

Comment:
Can we throw something else but not this exception?

Reply:
Fixed.